### PR TITLE
compress AST in memory, not just on disk

### PR DESCRIPTION
**Pull request by vtjnash**
sysimg size will be exactly the same, but seems to be pretty significant savings in memory and initial GC time (about 10% on each, estimated)

### CODE REVIEW

**Comment by ararslan** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

File: base/inference.jl [hunk 1]
```diff
@@ -3916,16 +3908,27 @@ function inlineable(f::ANY, ft::ANY, e::Expr, atypes::Vector{Any}, sv::Inference
         rettype = linfo.rettype
     end
 
+    inferred === nothing && return invoke_NF(argexprs0, e.typ, atypes, sv,
+                                        atype_unlimited, invoke_data)
```
Comment (ararslan) [inline base/inference.jl]: Kinda weird indentation here. `atype_unlimited` should probably be aligned beneath `argexprs0` for consistency with the `invoke_NF` call a few lines down.

**Comment by ararslan** [comment]
Looks like this did something weird in the inference code:
```
Error in testset inference:
Error During Test
  Test threw an exception of type ErrorException
  Expression: which(fpure, ()).source.pure
  type Array has no field pure
```

File: base/inference.jl [hunk 2]
```diff
@@ -3916,16 +3908,27 @@ function inlineable(f::ANY, ft::ANY, e::Expr, atypes::Vector{Any}, sv::Inference
...
         return invoke_NF(argexprs0, e.typ, atypes, sv, atype_unlimited,
                          invoke_data)
     end
 
+    if isa(inferred, CodeInfo)
+        ast = copy_exprargs(inferred.code)
+    else
+        ast = (ccall(:jl_uncompress_ast, Any, (Any, Any, Cint), method, inferred, true)::CodeInfo).code
+    end
+    ast = ast::Array{Any,1}
```
Comment (ararslan) [inline base/inference.jl]: I think this might be what's causing the issue with "`Array` has no field `code`."

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/20933#issuecomment-284850370) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/542b9fb37fc92e4b4d72785ad10630b1a0235e51/5e4ae7c_vs_8dc7064/report.md). cc @jrevels

**Comment by vtjnash** [comment]
@nanosoldier runbenchmarks("dates", vs=":master")

**Comment by pabloferz** [comment]
You missed some backticks.

@nanosoldier `runbenchmarks("dates", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/20933#issuecomment-285076073) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/798a250bd9842784d2c4c579f7dab83ea4e064d0/6477d13_vs_c29fc1e/report.md). cc @jrevels

**Comment by vtjnash** [comment]
Thanks.

OK, good. That confirms that the "dates"-"parse" benchmark was just noise (as I was seeing locally also).

File: src/dump.c [hunk 3]
```diff
@@ -2600,16 +2603,33 @@ JL_DLLEXPORT jl_array_t *jl_compress_ast(jl_method_t *m, jl_array_t *ast)
...
+        ios_write(s.s, namestr, namelen + 1); // include nul-byte
     }
 
-    //jl_printf(JL_STDERR, "%d bytes, %d values\n", dest.size, vals->length);
+    size_t nf = jl_datatype_nfields(jl_code_info_type);
+    for (i = 0; i < nf - 5; i++) {
+        jl_serialize_value_(&s, jl_get_nth_field((jl_value_t*)code, i), 1);
+    }
 
+    ios_putc('\0', s.s);
```
Comment (JeffBezanson) [inline src/dump.c]: Why is this needed?
